### PR TITLE
Refresh Cluster Provider API when extension are installed/uninstalled

### LINKED ISSUE #137: Refresh Cluster Provider API when extension are installed/uninstalled
Refresh Cluster Provider API when extensions are installed/uninstalled

### PR DESCRIPTION
**Pull request by angelozerr**
Refresh Cluster Provider API when extension are installed/uninstalled

Fixes #137

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by fbricon** [comment]
Works fine for displaying the custom cluster provider wizard (or not).
However, if you have an existing rhosak cluster and disabled vscode-rhoas, after re-enabling it, opening the topics fail to open because of `Cannot find cluster provider for 'rhosak' ID.`

**Comment by fbricon** [comment]
This is caused by caching the kafkaClient promise

https://github.com/jlandersen/vscode-kafka/blob/abd60c0c27bbdcc0a93725b28d8a76d08124324e/src/client/client.ts#L214-L230

**Comment by angelozerr** [comment]
> Works fine for displaying the custom cluster provider wizard (or not).
> However, if you have an existing rhosak cluster and disabled vscode-rhoas, after re-enabling it, opening the topics fail to open > because of Cannot find cluster provider for 'rhosak' ID.

I dispose the client accessor when extension changed.

File: src/extension.ts [hunk 1]
```diff
@@ -151,6 +151,13 @@ export function activate(context: vscode.ExtensionContext): KafkaExtensionPartic
             ConsumerVirtualTextDocumentProvider.SCHEME, consumerVirtualTextDocumentProvider)
     );
 
+    // Refresh cluster provider participant when a vscode extension is installed/uninstalled
+    if (vscode.extensions.onDidChange) {// Theia doesn't support this API yet
+        context.subscriptions.push(vscode.extensions.onDidChange(() => {
```
Comment (fbricon) [inline src/extension.ts:156]: only if relevant extensions (labeled kafka-provider) changed. No need to refresh the definitions if an unrelated extension was added/removed

File: src/kafka-extensions/registry.ts [hunk 2]
```diff
@@ -121,16 +122,22 @@ function sortByNameAscending(a: ClusterProvider, b: ClusterProvider): -1 | 0 | 1
...
 
 export interface ClusterProviderDefinition {
     id: string;
     name?: string;
 }
+export function refreshClusterProviderDefinitions() {
+    // collect cluster provider extensions
+    providers = collectClusterProviderDefinitions(vscode.extensions.all);
+    // Disconnect all clusters.
+    ClientAccessor.getInstance().dispose();
```
Comment (fbricon) [inline src/kafka-extensions/registry.ts:139]: only if providers list changed

File: src/extension.ts (hunk 1 above)
Comment (angelozerr) [reply src/extension.ts:156]: Refresh a definition is just loop for package.json and get information about cluster provider. I have not found the caability to know which extension is installing, so you need to loop for all extensions and search in package.json the labeled kafka-provider. It's the same thing than collect cluster provider information.

File: src/kafka-extensions/registry.ts (hunk 2 above)
Comment (angelozerr) [reply src/kafka-extensions/registry.ts]: Now I dispose clients which have a cluster provider defined before/after installed extensions

File: src/kafka-extensions/registry.ts [hunk 3]
```diff
@@ -121,16 +122,28 @@ function sortByNameAscending(a: ClusterProvider, b: ClusterProvider): -1 | 0 | 1
...
+export function refreshClusterProviderDefinitions() {
+    // Collect cluster provider extensions
+    const oldClusterProviderIds = Array.from(providers.keys());
+    providers = collectClusterProviderDefinitions(vscode.extensions.all);
+
+    // Disconnect all kafka client linked to a cluster provider id coming from an installed/uninstalled extension
+    const newClusterProviderIds = Array.from(providers.keys());
+    const allClusterProviderIds = [...oldClusterProviderIds, ...newClusterProviderIds]
+        .filter(clusterProviderId => clusterProviderId !== defaultClusterProviderId);
+    const toDispose = [...new Set(allClusterProviderIds)];
```
Comment (fbricon) [inline src/kafka-extensions/registry.ts:144]: no, you should only dispose old is that are not in the new newClusterProviderIds.
Something like:
```
idsToDispose = oldClusterProviderIds.filter( id => !newClusterProviderIds.includes(id) )
```

Comment (angelozerr) [reply src/kafka-extensions/registry.ts]: Thanks for your suggestion, I tried on my side an dit works great. Fixed.

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ All notable changes to `Tools for Apache Kafka®` are documented in this file.
 - Simplify snippets. See [#180](https://github.com/jlandersen/vscode-kafka/pull/180).
 - Hover support in `.kafka` files. See [#149](https://github.com/jlandersen/vscode-kafka/issues/149).
 - String encoding serialization support. See [#181](https://github.com/jlandersen/vscode-kafka/issues/181).
+- Refresh Cluster Provider API when extension are installed/uninstalled. See [#137](https://github.com/jlandersen/vscode-kafka/issues/137).
```
Comment (fbricon) [inline CHANGELOG.md:12]: extensions are

File: src/kafka-extensions/registry.ts [hunk 5]
```diff
@@ -121,16 +122,29 @@ function sortByNameAscending(a: ClusterProvider, b: ClusterProvider): -1 | 0 | 1
...
+    providers = collectClusterProviderDefinitions(vscode.extensions.all);
+    const newClusterProviderIds = Array.from(providers.keys());
+
+    // Disconnect all kafka client linked to a cluster provider id coming from an installed/uninstalled extension
+    const oldIdsToDispose = oldClusterProviderIds.filter(id => !newClusterProviderIds.includes(id));
+    const newIdsToDispose = newClusterProviderIds.filter(id => !oldClusterProviderIds.includes(id));
+    const allIdstoDispose = [...oldIdsToDispose, ...newIdsToDispose];
+    const toDispose = [...new Set(allIdstoDispose)];
+
+    ClientAccessor.getInstance().dispose(toDispose);
```
Comment (fbricon) [inline src/kafka-extensions/registry.ts:146]: const allIdsToDispose = ...
if (allIdsToDispose.length > 0) {
const toDispose = [...new Set(allIdstoDispose)];
ClientAccessor.getInstance().dispose(toDispose);
}

File: CHANGELOG.md (hunk 4 above)
Comment (angelozerr) [reply CHANGELOG.md]: fixed.

File: src/kafka-extensions/registry.ts (hunk 5 above)
Comment (angelozerr) [reply src/kafka-extensions/registry.ts]: fixed.

**Comment by fbricon** [comment]
Thanks @angelozerr

**Pull request merged**